### PR TITLE
As a User, I'd like to be able to have a customized Profile.

### PR DESCRIPTION
**Pull request by AlexLLKong**
This is the user's profile page. They can edit their name, email, and password.

image
Figma wireframe

- [x] HTML Components
    - [x] Profile
    - [x] Footer

- [x] Endpoints
    - [x] `PATCH /profile`

### CODE REVIEW

**Comment by nok-ko** [comment]
@AlexLLKong just changed the `base` branch on the PR to `dev`. 😅 It's easy to forget.

File: main.js [hunk 1]
```diff
@@ -251,13 +402,15 @@ app.post("/signup", async (req, res) => {
...
 	// Set response header regardless of success/failure
 	res.setHeader("Content-Type", "application/json");
 
 	try {
-		await db.collection('BBY-6_users')
-			.insertOne({name, emailAddress, pwd, avatarURL, date, achievements, admin, kit});
+		bcrypt.hashSync(req.body.password, 10, async(err, pwd) => {
+			await db.collection('BBY-6_users')
+				.insertOne({name, emailAddress, pwd, avatarURL, date, achievements, admin, kit});
+		});
```
Comment (nok-ko) [inline main.js:413]: Can use `await bcrypt.hash(…)` here instead of `.hashSync(…)`.

File: main.js [hunk 2]
```diff
@@ -165,27 +193,149 @@ app.use("/images", express.static("public/images"));
...
+	})
+	res.status(200).send("avatarUploaded");
+})
+
 app.get('/login', function (req, res) {
+	// Redirect to profile page if logged in
+	if (req.session.loggedIn) {
+		res.redirect("/profile");
+		return;
+	}
```
Comment (nok-ko) [inline main.js:334]: Closes login-redirecting part of #19.

File: main.js [hunk 3]
```diff
@@ -165,27 +193,149 @@ app.use("/images", express.static("public/images"));
...
+
+	// if the request is not coming from a logged in user, reject.
+	if (!req.session || !req.session.loggedIn) {
+		res.status(401).send("invalidSession");
+		return;
+	} 
+	
+	// get user from db
+	const filterQuery = {emailAddress: req.session.email};
+	const userResults = await db.collection('BBY-6_users').find(filterQuery).toArray();
```
Comment (nok-ko) [inline main.js:248]: Wonder if it's better to skip this, and just do a single request to the database, catch this error case on the update failing. Reasoning being that two requests are slower than one, and that there's actually a weird race condition here, if someone deletes a user you want to update, just after you run the "user exists" query.

File: main.js [hunk 4]
```diff
@@ -165,27 +193,149 @@ app.use("/images", express.static("public/images"));
...
+
+app.post('/avatar', upload.single("avatar"), async(req, res) => {
+	//if the request is not coming from a logged in user, reject.
+	if (!req.session || !req.session.loggedIn) {
+		res.status(401).send("invalidSession");
+		return;
+	} 
+	const results = await db.collection('BBY-6_users').updateOne({emailAddress:req.session.email}, {
+		$set: {
+			avatarURL: {
```
Comment (nok-ko) [inline main.js:320]: Probably shouldn't call this `avatarURL` any more.

File: public/css/index.css [hunk 5]
```diff
@@ -119,3 +125,15 @@ body {
 .fail {
 	color: red;
 }
+
+#Footer-Links {
```
Comment (nok-ko) [inline public/css/index.css:129]: We should make sure all of our CSS follows the same naming convention sometime.

File: public/css/signup-success-instruction.css [hunk 6]
```diff
@@ -7,17 +7,20 @@ body{
...
     justify-items: center;
-    padding-top: 10vh;
-    margin: 10vw;
+    padding-top: 50px;
+    margin: 50px;
     display: grid;
     grid-template-columns: 1fr;
     grid-template-rows: repeat(5, auto);
-    font-size: 4vw;
+    font-size: 25px;
```
Comment (nok-ko) [inline public/css/signup-success-instruction.css:15]: Use `rem` for font sizes!

File: public/css/user-profile.css [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+    -ms-user-select: none;  
+    -webkit-user-select: none;   
+    background-color: #FFFFFF;  
+    box-shadow: 1.5px 1.5px 2px rgb(202, 187, 144);
+    border-radius: 25px;
+    padding: 5px;
+}
+
+#Pen-1:hover, #Pen-2:hover, #Pen-3:hover, #Pen-4:hover{
+    color: #c7c7c7;
```
Comment (nok-ko) [inline public/css/user-profile.css:104]: Put some commonly-used colours in CSS custom properties like:

```css
:root {
    --color-info: #c7c7c7;
}
p {
    color: var(--color-info);
} 
```
Makes it easier to keep track of the palette.

File: public/css/user-profile.css [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+
+#Base-Container-BackToKit:hover > span, #Base-Container-BackToKit:hover > p{
+    color: #2dc7df;
+}
+
+#Update-Error{
+    padding-bottom: 10px;
+}
+
+::-webkit-file-upload-button {
```
Comment (nok-ko) [inline public/css/user-profile.css:248]: WebKit-specific selector... Have you tested on browsers other than Chrome? 😅

File: public/js/client.js [hunk 9]
```diff
@@ -9,33 +9,38 @@ function ready(callback) {
...
+    const password = document.getElementById("Password").value;
+    const response = await fetch("/login", {
+        headers: {
+            'Content-Type': 'application/json',
+        },
+        "method": "POST",
+        "body": JSON.stringify({email, password})
+    });
+    
+    // alexkonghockey@gmail.com
```
Comment (nok-ko) [inline public/js/client.js:25]: Leftover comments...

```suggestion
```

File: public/js/client.js [hunk 10]
```diff
@@ -9,33 +9,38 @@ function ready(callback) {
...
+    // not_bacon
```
Comment (nok-ko) [inline public/js/client.js:26]:
```suggestion
```

File: public/js/client.js [hunk 11]
```diff
@@ -9,33 +9,38 @@ function ready(callback) {
...
+				document.getElementById("Error-Msg").innerText = "Invalid Email/Password";
+				break;
             default:
                 break;
         }
 
-    });
+});
 
+ready(function(){
```
Comment (nok-ko) [inline public/js/client.js:44]: What's this ready business about?

File: public/js/user-profile.js [hunk 12]
```diff
@@ -0,0 +1,181 @@
+"use strict";
+
+const UpdateButton = document.getElementById("Update-Button");
```
Comment (nok-ko) [inline public/js/user-profile.js:3]: In JS, PascalCase identifiers are usually reserved for class names.

File: public/js/user-profile.js [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+		let base64 = responseJson.data;
+		base64 = `data:${responseJson.mimeType};base64,${base64}`;
+		document.getElementById("Base-Container-ProfilePicture").src = base64;
+	} catch(e) {
+		console.log(e);
+	}
+}
+
+async function executeUpdate(){
+	checkForInvalidInput(FullNameInput);
```
Comment (nok-ko) [inline public/js/user-profile.js:46]: Could shorten this to:

```js
[fooInput, barInput, bazInput].map(checkForInvalidInput);
```

File: public/js/user-profile.js [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+        PasswordInput.disabled= false; 
+        PasswordInput.focus();
+        PasswordInput.onfocus = clearInput(PasswordInput);
+    } else {
+        PasswordInput.disabled= true; 
+        PasswordInput.value = "12345" 
+    }
+}
+
+function checkForInvalidInput(input) {
```
Comment (nok-ko) [inline public/js/user-profile.js:147]: Better name might be `validate`.

File: public/js/user-profile.js [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+
+	getAvatar();
+	getProfileDetails();
+}
+
+function clearInput(x){
+    x.value = '';
+}
+
+function editFullName(){
```
Comment (nok-ko) [inline public/js/user-profile.js:116]: These `editFoo` functions could be generated by a higher order function, don't repeat yourself.

File: public/js/user-profile.js [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+
+    const response = await fetch("/profile", {
+        method: "PATCH",
+        headers: {'Content-Type':'application/json'},
+        body: JSON.stringify(payload)
+    })
+
+    const ResponseText = await response.text(); 
+	const feedback = document.getElementById("Update-Error");
+    switch(ResponseText){
```
Comment (nok-ko) [inline public/js/user-profile.js:90]: Could replace the switch... case with an object mapping error IDs to strings/actions.

File: main.js [hunk 17]
```diff
@@ -141,6 +143,32 @@ const initDatabase = async(db) => {
...
+	baseDOM = loadHTMLComponent(baseDOM, "header", "header", "./html/header.html");
+
+	// Add the footer
+	baseDOM = loadHTMLComponent(baseDOM, "footer", "footer", "./html/footer.html");
+
+	return baseDOM;
+}
+
+// Load a HTML component 
+// baseDOM is the return value of `new JSDOM(htmlFile)`
```
Comment (nok-ko) [inline main.js:159]: Just use JSDoc comment style.

**Review by nok-ko — COMMENTED** [review]
Review made on my phone, pardon some of the curt comments (and ask clarification if you need!)

Will come back with another review once I'm at a computer.

**Comment by nok-ko** [comment]
Also please try to make sure all of your functions have header comments describing what they do, ideally in JSDoc format

File: public/js/user-profile.js [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+            break;
+        case "emailInUse":
+            // Email already in use.
+			feedback.innerText = "That email is already in use";
+			feedback.style.color = "red";
+            break;
+        default:
+            feedback.innerText = "There was an issue on the server";
+			feedback.style.color = "red";
+    }
```
Comment (nok-ko) [inline public/js/user-profile.js:104]: What I meant:

```suggestion
    const ResponseText = await response.text(); 
    const feedbackEl = document.getElementById("Update-Error");
    const [newText, newStyle] = {
        "userUpdated": ["Successfully updated profile.", { color: 'green' }],
        "emailInUse": ["That email is already in use.", { color: 'red' }],
        "_default": ["There was an issue on the server.", { color: 'red' }]
    }[ResponseText ?? '_default'];
    
    feedback.textContent = newText;
    Object.assign(feedback.style, newStyle);
```

And the even better thing to do would be to put that error message style (`color: red`/`color: green`) into a CSS class so that we can edit it from the stylesheet and not have too much markup in the JavaScript.

File: public/js/user-profile.js (hunk 15 above)
Comment (nok-ko) [reply public/js/user-profile.js:116]: E.g:

```js
function makeEditAction(inputEl, defaultValue){
    return function () {
        if (inputEl.disabled){
            inputEl.disabled= false; 
            inputEl.focus();
            inputEl.onfocus = clearInput(inputEl);
            return;
        }
        inputEl.disabled = true;
        inputEl.value = defaultValue;
    }
}

const editEmail = makeEditAction(EmailInput, userEmail);
const editFullName = makeEditAction(FullNameInput, userName);
const editPassword = makeEditAction(PasswordInput, "12345");
```

File: public/js/user-profile.js [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+	const file = this.files[0];
+	const error = document.getElementById("Avatar-Error");
+	if(file && file.size > 8 * 1024 * 1024) {
+		// Don't allow files larger than 8MB
+		error.innerText = "Max file size is 8MB";
+		errors = errors | errorFlags.Avatar;
+	} else {
+		error.innerText = "";
+		errors = errors & ~(errorFlags.Avatar);
+	}
```
Comment (nok-ko) [inline public/js/user-profile.js:171]: Should have server-side validation for this, too.

**Review by nok-ko — CHANGES_REQUESTED** [review]
More substantial review. General recommendations aside from the specific stuff:

- Look for reused or “magic” CSS values, put them in CSS custom properties.
- Add doc comments to any functions longer than a couple of lines. (Not a big priority, but nice going forward.)
- Avoid `readFileSync`, use the functions in the `node:fs/promises` package instead.

Other side stuff:
- Default profile picture could be an SVG instead of what we have now. Okay for now, since it's just a placeholder.
- Consider putting some of the clientside JS in separate ES6 modules, especially stuff we might re-use.
- I don't quite understand how the error flag system works, so can't really offer cogent criticism… It seems to make sense, but I'm not sure.

Otherwise, I haven't looked at the actual UX of the thing, just the code… something to do later. 😅

File: main.js [hunk 20]
```diff
@@ -165,27 +193,149 @@ app.use("/images", express.static("public/images"));
...
+		// name to be changed
+		updateQuery['name'] = req.body.name;
+	} 
+	if (req.body.pwd) {
+		// password to be changed
+		const pwd = await bcrypt.hash(req.body.pwd, 10);	
+		updateQuery['pwd'] = pwd;				
+	} 
+	 
+	if(updateQuery === {}){
```
[the remark this answers is not in the document]
Comment (nok-ko) [reply main.js:271]: Not a very important error (just nice feedback to the user of the API,) but would be nice if it was handled. Can leave it the way it is for now, maybe just make an issue about it to fix later.

File: main.js (hunk 3 above)
Comment (nok-ko) [reply main.js:248]: Tracked in #37, resolving this for now.

File: main.js (hunk 4 above)
Comment (nok-ko) [reply main.js:320]: Trackedin #36, resolving for now.

File: html/user-profile.html [hunk 21]
```diff
@@ -0,0 +1,38 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <title>ShakeGuard - Profile</title>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <link href="images/comp2800_logo_favicon.ico" rel="icon" type="image/x-icon" />
+    <link href="css/user-profile.css" rel="stylesheet"/>
```
[the remark this answers is not in the document]
Comment (nok-ko) [reply html/user-profile.html:8]: Alex says they're not used anymore – will remove.

File: main.js [hunk 22]
```diff
@@ -165,27 +193,149 @@ app.use("/images", express.static("public/images"));
...
+		// invalid body, reject request
+		res.status(400).send("missingBodyArgument(s)");
```
[the remark this answers is not in the document]
Comment (nok-ko) [reply main.js:273]: Client handles it, so the API is complete enough.

File: main.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (nok-ko) [reply main.js:413]: Actually, if we don't use the `/signup` endpoint, we should just remove it entirely.

File: public/css/user-profile.css [hunk 23]
```diff
@@ -0,0 +1,263 @@
...
+    grid-row: 3 / span 1;
+    margin-left: 1em;
+}
+
+#Base-Container-Upload-Avatar-Centering-Container{
+    display: grid;
+    grid-template-columns: 1fr;
+    grid-template-rows: 1fr;
+    grid-column: 1 / span 1;
+    grid-row: 2 / span 1;
```
[the remark this answers is not in the document]
Comment (nok-ko) [reply public/css/user-profile.css:52]: Minor thing, just consider doing it later. Won't fix, it's fine.

File: html/profile-component.html [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+    <input id="Upload-Avatar" type="file" name="Upload-Avatar" accept=".jpg,.png,.jpeg"/>
+  </div>
+
+  <!-- Full Name -->
+  <div id="Base-Container-FullName-Container">
+    <form class="Container-Form">
+      <input disabled="disabled" type="text" id="FullName" class="Form-Input" value="FirstName LastName"/>
+      <span id="Pen-1" class="material-symbols-outlined">edit</span>
+      <p class="error" id="FullName-Error"></p>
+    </form>
```
[the remark this answers is not in the document]
Comment (nok-ko) [reply html/profile-component.html:30]: Fine for now, but should resolve this.

File: public/js/user-profile.js (hunk 19 above)
Comment (nok-ko) [reply public/js/user-profile.js:171]: Don't care about this for the prototype.

**Comment by nok-ko** [comment]
Workflow runs are failing because templates are not masked from the HTML Validator. Sigh, will fix later.

**Comment by AlexLLKong** [comment]
## Post Code Review Changes
- [x] Move html components into separate folder and naming convention.
- [x] Credentials in the README.md for reg user and admin.
- [x] Strip out ready function.
- [x] Strip out profile.html and profile.css.
- [x] Use await for fs.
- [x] Add session check for /profile-details.
    - Also refactor frequently-used redirect/error check code into functions: `checkLoginError` and `redirectToLogin`.
    - Should really be Express middlewares.
- [x] Strip out /signup.
- [x] Check for console.logs.
- [x] Fix stretching on profile pictures

**Comment by nok-ko** [comment]
[Code review with Alex](https://trello.com/c/jzCKGE2B/190-code-review-with-alex)

**Review by nok-ko — APPROVED** [review]
Wow, even made the checks pass! Awesome work, guys.

**Comment by AlexLLKong** [comment]
> Wow, even made the checks pass! Awesome work, guys.
wait why did the checks pass? I don't think anything was done to make them pass

**Comment by nok-ko** [comment]
> wait why did the checks pass? I don't think anything was done to make them pass

'Cause the workflow was only set up to check files in `html/` :)

**Comment by nok-ko** [comment]
[Code review and merge branches (1hr) (1) [4]](https://trello.com/c/kh3fiDMq/178-code-review-and-merge-branches-1hr-1-4)
